### PR TITLE
improve fetch USD price

### PR DESCRIPTION
**Pull request by NNagato**

### CODE REVIEW

File: lib/tokenrate/cached_rate_provider.go [hunk 1]
```diff
@@ -8,20 +8,42 @@ import (
...
-func NewCachedRateProvider(sugar *zap.SugaredLogger, provider tokenrate.ETHUSDRateProvider, timeout time.Duration) *CachedRateProvider {
-	return &CachedRateProvider{
+func NewCachedRateProvider(sugar *zap.SugaredLogger, provider tokenrate.ETHUSDRateProvider, options ...CachedRateProviderOption) *CachedRateProvider {
+	crp := &CachedRateProvider{
 		sugar:    sugar,
-		timeout:  timeout,
 		provider: provider,
 	}
+	for _, option := range options {
+		option(crp)
```
Comment (favadi) [inline lib/tokenrate/cached_rate_provider.go]: what will happen if user provides no timeout option? We should either handle it or provide a default value.

File: lib/tokenrate/cached_rate_provider.go [hunk 2]
```diff
@@ -8,20 +8,42 @@ import (
...
 
 //CachedRateProvider is cached provider for eth usd rate
 type CachedRateProvider struct {
-	sugar    *zap.SugaredLogger
-	timeout  time.Duration
-	provider tokenrate.ETHUSDRateProvider
+	sugar       *zap.SugaredLogger
+	timeout     time.Duration
+	provider    tokenrate.ETHUSDRateProvider
+	withWarning bool
```
Comment (favadi) [inline lib/tokenrate/cached_rate_provider.go]: withWarning should be called warningOnly.

Comment (NNagato) [reply lib/tokenrate/cached_rate_provider.go]: fixed in https://github.com/KyberNetwork/reserve-stats/pull/330/commits/78ada81cd8816f253aef998174ebdf909b84d0c8

File: lib/tokenrate/cached_rate_provider.go (hunk 1 above)
Comment (NNagato) [reply lib/tokenrate/cached_rate_provider.go]: I think we will provide a default value for timeout, fixed in https://github.com/KyberNetwork/reserve-stats/pull/330/commits/5976ca16a471a330477a99770db29f22e7543c39

**Pull request merged**